### PR TITLE
automagically detect the length of the iteration data

### LINKED ISSUE #534: Dynamic iteration through data file no longer works?
Hi,

Previously I used v2 newman, and I'm sure when passing in data files, I used -n which iterated through all the rows. Using v3 now, it requires a number e.g. -n 2. Can we fix this?

Regards,

Aaron

### PR DESCRIPTION
**Pull request by czardoz**
Fixes #534

### CODE REVIEW

**Comment by shamasis** [comment]
I like! Sadly all build failing! 🙄

**Comment by kunagpal** [comment]
Not anymore.

File: lib/run/index.js [hunk 1]
```diff
@@ -82,7 +82,8 @@ module.exports = function (options, callback) {
         runner.run(options.collection, {
             stopOnFailure: options.bail, // LOL, you just got trolled ¯\_(ツ)_/¯
             abortOnFailure: options.abortOnFailure,
-            iterationCount: options.iterationCount,
+            iterationCount: options.iterationCount ||
+                (_.get(options, 'iterationData') ? options.iterationData.length : 1),
```
Comment (shamasis) [inline lib/run/index.js]: Will this not error out if iteration data is undefined or not an array? @czardoz

Comment (shamasis) [inline lib/run/index.js]: Ignore above comment. 

Why not directly do _.get(options, "iterationData.length") || 1

Comment (czardoz) [inline lib/run/index.js]: If iteration data is present, we should not pass any iteration count.

**Comment by czardoz** [comment]
@kunagpal Can you add a small CLI level test? 

If we run Newman like this:
 
```
newman run collection --iteration-data ./my-csv.csv
```

By default, the number of iterations should equal the number of rows in the CSV.

File: lib/run/index.js (hunk 1 above)
Comment (kunagpal) [inline lib/run/index.js]: @czardoz So the intended order of precedence is:

1. iterationData.length
2. iterationCount
3. 1

**Comment by kunagpal** [comment]
Alright.

File: lib/run/index.js [hunk 2]
```diff
@@ -82,7 +82,7 @@ module.exports = function (options, callback) {
         runner.run(options.collection, {
             stopOnFailure: options.bail, // LOL, you just got trolled ¯\_(ツ)_/¯
             abortOnFailure: options.abortOnFailure,
-            iterationCount: options.iterationCount || _.get(options, 'iterationData.length') || 1,
+            iterationCount: _.get(options, 'iterationData.length') || options.iterationCount || 1,
```
Comment (czardoz) [inline lib/run/index.js]: If iteration count is specified, it should override the length of iteration data.

File: test/unit/run.test.js [hunk 3]
```diff
@@ -70,4 +71,58 @@ describe('run module', function () {
             }
         ], done);
     });
+
+    it('must correctly resolve conflicts between iterationData.length and iterationCount', function (done) {
```
Comment (czardoz) [inline test/unit/run.test.js]: By CLI level tests, I meant these: https://github.com/postmanlabs/newman/blob/develop/test/cli/suppress-exit-code.test.js

You should add a spec like that, with iteration data as a CSV file, and then run it in these combinations:
1. `--iteration-data` absent and `--iteration-count` absent (should run one iteration)
2. `--iteration-data` present and `--iteration-count` absent (should run iterations equal to number of rows in iteration data)
3. `--iteration-data` absent and `--iteration-count` present (should run iterations equal to iteration count given)
4. `--iteration-data` present and `--iteration-count` present (should always run iterations equal to given iteration count)

File: lib/run/index.js [hunk 4]
```diff
@@ -82,7 +82,7 @@ module.exports = function (options, callback) {
         runner.run(options.collection, {
             stopOnFailure: options.bail, // LOL, you just got trolled ¯\_(ツ)_/¯
             abortOnFailure: options.abortOnFailure,
-            iterationCount: options.iterationCount,
+            iterationCount: _.get(options, 'iterationData.length') || options.iterationCount || 1,
```
Comment (czardoz) [inline lib/run/index.js]: iteration count should always override the length of iteration data

File: test/unit/run.test.js (hunk 3 above)
Comment (kunagpal) [inline test/unit/run.test.js]: Okay, how do we verify the number of iterations that occurred from within an `exec` call?

Comment (shamasis) [inline test/unit/run.test.js]: By using nock to mock the API and on its reply add callback function with expect. 
https://www.npmjs.com/package/nock

But for now, simply add that it shouldn't throw non zero code and add a todo

File: lib/run/index.js [hunk 5]
```diff
@@ -82,7 +82,7 @@ module.exports = function (options, callback) {
         runner.run(options.collection, {
             stopOnFailure: options.bail, // LOL, you just got trolled ¯\_(ツ)_/¯
             abortOnFailure: options.abortOnFailure,
-            iterationCount: options.iterationCount,
+            iterationCount: options.iterationCount || _.get(options, 'iterationData.length'),
```
Comment (shamasis) [inline lib/run/index.js]: @czardoz just a thought - if runtime already handles this, we shouldn't even try parsing anything from iteration data.

**Pull request merged**